### PR TITLE
Fixing Bag memory problem with Swift 2.1 (fixes #2417)

### LINKED ISSUE #2417: Xcode 7.1-beta2: Signal observe crash with multiple observers
Following code crashes after some iterations: 

```
let signal: Signal<Void, NoError> = Signal { _ in
    return nil
}

while(true) {
   let disposable = signal.observeCompleted {}
   disposable?.dispose()
}
```

Console log:
```
Test(78365,0x107691300) malloc: *** mach_vm_map(size=70918499991552) failed (error code=3)
*** error: can't allocate region
*** set a breakpoint in malloc_error_break to debug
```

It crashes on this line: https://github.com/ReactiveCocoa/ReactiveCocoa/blob/v4.0-alpha.1/ReactiveCocoa/Swift/Bag.swift#L35

Is this a Swift or ReactiveCocoa bug or am I doing something wrong?

### PR DESCRIPTION
**Pull request by NachoSoto**
http://www.openradar.me/22873620

This problem has been fixed in beta 3, but I've added test coverage to verify that it never breaks again.

### Changes:
- Re-added `BagSpec.swift` to the suite (oops)
- Added a test reproducing the problem.

### CODE REVIEW

File: ReactiveCocoaTests/Swift/SignalSpec.swift [hunk 1]
```diff
@@ -307,6 +307,20 @@ class SignalSpec: QuickSpec {
...
 				expect(completed).to(beTruthy())
 			}
+
+			context("memory") {
+				it("should not crash allocating memory with a few observers") {
+					let (signal, sink) = Signal<Int, NoError>.pipe()
+
+					for i in 0..<50 {
+						autoreleasepool {
+							let disposable = signal.observe { _ in }
```
Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalSpec.swift]: This crashes in the 39th iteration **only** in Swift 2.1 (see [radar](http://www.openradar.me/22873620)) before the workaround.

**Comment by NachoSoto** [comment]
Hmm looks like this workaround isn't very valid, my app is deadlocking sometimes (only on a device).

**Comment by NachoSoto** [comment]
Finally a real fix! :D This is ready for review now.

**Comment by NachoSoto** [comment]
Okay, this fixes the issue with `Bag`... but now my app is crashing with the same issue in some other `Array.append` :-1: so maybe it's not worth landing this, we'd have to replace every `Array.append` with `array + [newElement]`...

What does everyone think? Should we just merge the other 2 changes (the test and re-adding `BagSpec.swift`)?

**Comment by dapenggao** [comment]
Probably we should just stick to the current version right now. When 2.1 becomes stable we can easily migrate to it without any pain.

**Comment by NachoSoto** [comment]
> Probably we should just stick to the current version right now. When 2.1 becomes stable we can easily migrate to it without any pain.

Agreed.

**Comment by NachoSoto** [comment]
All right I reverted the workaround since that's no longer necessary and beta 3.
I'm merging this since it's only test coverage but no functional change :hammer:

**Comment by craigzour** [comment]
Hi guys,

I received a crash report during the weekend (on Crashlytics). I think it is related to this thread because it points to Bag.swift and especially this instruction "Array.append(A) -> ()".

I am using Reactive Cocoa 4.2.0.

I did not put the complete stack trace because it is huge but here a few lines.

0   libsystem_malloc.dylib         0x18315d4a0 szone_malloc_should_clear + 6
1   libsystem_malloc.dylib         0x18315d458 malloc_zone_malloc + 116
2   libsystem_malloc.dylib         0x1831619dc malloc + 52
3   libswiftCore.dylib             0x1015f008c swift_slowAlloc (__hidden#8343_:27)
4   libswiftCore.dylib             0x1015f00dc __hidden#8345_ (__hidden#8365_:90)
5   libswiftCore.dylib             0x10143bc90 _ContiguousArrayBuffer.init(count : Int, minimumCapacity : Int) -> _ContiguousArrayBuffer<A> (__hidden#4857_:200)
6   libswiftCore.dylib             0x1014477b8 _forceCreateUniqueMutableBuffer<A where ...> (inout A, countForNewBuffer : Int, minNewCapacity : Int) -> _ContiguousArrayBuffer<A.Element> (__hidden#4857_:2424)
7   libswiftCore.dylib             0x10145d058 Array._copyToNewBuffer(Int) -> () (__hidden#4857_:1871)
8   ReactiveCocoa                  0x100e4a7b4 Array.append(A) -> () (Bag.swift)
9   ReactiveCocoa                  0x100e1e48c (ConcatState in _F9B5D3FB407CAF9A5D2D355C6E47C3CE).(enqueueSignalProducer(SignalProducer<A, B>) -> ()).(closure #1) (Flatten.swift)
10  ReactiveCocoa                  0x100e269ec partial apply for thunk (Flatten.swift)
11  ReactiveCocoa                  0x100e4374c Atomic.(modify((A) throws -> A) throws -> A).(closure #1) (Atomic.swift:61)
12  ReactiveCocoa                  0x100e43338 Atomic.withValue<A> ((A) throws -> A1) throws -> A1 (Atomic.swift:74)
13  ReactiveCocoa                  0x100e4326c Atomic.modify((A) throws -> A) throws -> A (Atomic.swift:63)
14  ReactiveCocoa                  0x100e1dc28 (ConcatState in _F9B5D3FB407CAF9A5D2D355C6E47C3CE).enqueueSignalProducer(SignalProducer<A, B>) -> () (Flatten.swift:437)
15  ReactiveCocoa                  0x100e1dac0 SignalType<A where ...>.((observeConcat in _F9B5D3FB407CAF9A5D2D355C6E47C3CE)(Observer<A.Value.Value, A.Error>, CompositeDisposable?) -> Disposable?).(closure #1) (Flatten.swift:343)
16  ReactiveCocoa                  0x100db6880 Signal.(init((Observer<A, B>) -> Disposable?) -> Signal<A, B>).(closure #1) (Signal.swift:64)
17  ReactiveCocoa                  0x100df7eb8 SignalProducer.(startWithSignal((Signal<A, B>, Disposable) -> ()) -> ()).(closure #2) (SignalProducer.swift:255)
18  ReactiveCocoa                  0x100db6880 Signal.(init((Observer<A, B>) -> Disposable?) -> Signal<A, B>).(closure #1) (Signal.swift:64)
19  ReactiveCocoa                  0x100dba4cc SignalType.(map<A> ((A.Value) -> A1) -> Signal<A1, A.Error>).(closure #1).(closure #1) (Signal.swift)
20  ReactiveCocoa                  0x100deb1f0 partial apply for SignalType.(map<A> ((A.Value) -> A1) -> Signal<A1, A.Error>).(closure #1).(closure #1) (Signal.swift)

**Comment by andersio** [comment]
Do you have the error message and/or the error code?

**Comment by craigzour** [comment]
The only thing I can add as an information is "EXC_BAD_ACCESS KERN_PROTECTION_FAILURE 0x000000016e47ffe0"

**Comment by andersio** [comment]
Do you see a repetitive occurrence of `ConcatState` in the long stack trace? If you do, it is probably due to #3140, which has been fixed in https://github.com/ReactiveCocoa/ReactiveSwift/pull/5.

**Comment by craigzour** [comment]
@andersio Yes I have a lot of them. Thanks for your help ! I will have to wait for the migration to Swift 3 in order to get this fix.

**Pull request merged**